spring-cloud-stream-integration-tests/src/test/java/org/springframework/cloud/stream/config/StreamListenerWithConditionsTest.java

```diff
@@ -0,0 +1,120 @@
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.handler.annotation.Payload;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+/**
+ * @author Marius Bogoevici
+ */
+public class StreamListenerWithConditionsTest {
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void testAnnotatedArgumentsWithConditionalClass() throws Exception {
+		ConfigurableApplicationContext context = SpringApplication.run(TestPojoWithAnnotatedArguments.class,
+				"--server.port=0");
+
+		TestPojoWithAnnotatedArguments testPojoWithAnnotatedArguments = context
+				.getBean(TestPojoWithAnnotatedArguments.class);
+		Sink sink = context.getBean(Sink.class);
+		String id = UUID.randomUUID().toString();
+		sink.input().send(MessageBuilder.withPayload("{\"foo\":\"barbar" + id + "\"}")
+				.setHeader("contentType", "application/json").setHeader("testHeader", "testValue")
+				.setHeader("type", "foo").build());
+		sink.input().send(MessageBuilder.withPayload("{\"bar\":\"foofoo" + id + "\"}")
+				.setHeader("contentType", "application/json").setHeader("testHeader", "testValue")
+				.setHeader("type", "bar").build());
+		sink.input().send(MessageBuilder.withPayload("{\"bar\":\"foofoo" + id + "\"}")
+				.setHeader("contentType", "application/json").setHeader("testHeader", "testValue")
+				.setHeader("type", "qux").build());
+		assertThat(testPojoWithAnnotatedArguments.receivedFoo).hasSize(1);
+		assertThat(testPojoWithAnnotatedArguments.receivedFoo.get(0)).hasFieldOrPropertyWithValue("foo",
+				"barbar" + id);
+		assertThat(testPojoWithAnnotatedArguments.receivedBar).hasSize(1);
+		assertThat(testPojoWithAnnotatedArguments.receivedBar.get(0)).hasFieldOrPropertyWithValue("bar",
+				"foofoo" + id);
+		context.close();
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void testConditionalFailsWithDeclarativeMethod() throws Exception {
+		try {
+			ConfigurableApplicationContext context = SpringApplication.run(TestConditionalOnDeclarativeMethodFails.class,
+					"--server.port=0");
+			context.close();
+			fail("Context creation failure expected");
+		} catch (BeanCreationException e) {
+			assertThat(e).hasRootCauseInstanceOf(IllegalArgumentException.class);
+			assertThat(e.getCause()).hasMessageContaining("Cannot set a condition when using @StreamListener in declarative mode");
+		}
+	}
+
+	@EnableBinding(Sink.class)
+	@EnableAutoConfiguration
+	public static class TestPojoWithAnnotatedArguments {
+
+		List<StreamListenerTestUtils.FooPojo> receivedFoo = new ArrayList<>();
+
+		List<StreamListenerTestUtils.BarPojo> receivedBar = new ArrayList<>();
+
+		@StreamListener(value = Sink.INPUT, condition = "headers['type']=='foo'")
+		public void receive(@Payload StreamListenerTestUtils.FooPojo fooPojo) {
```

Comment:
Updated how we handle multiple return values - a return value is an error condition now.